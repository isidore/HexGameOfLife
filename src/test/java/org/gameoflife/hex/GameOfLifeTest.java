package org.gameoflife.hex;

import org.approvaltests.Approvals;
import org.junit.jupiter.api.Test;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class GameOfLifeTest {


    @Test
    void testZeroDies() {
        String timeline = "";
        GameOfLife gameOfLife = new GameOfLife();
        gameOfLife.board.setAlive(4, 4);
        timeline += gameOfLife;
        GameOfLife gameOfLife2 = gameOfLife.advanceTurn();
        timeline += "\n NEXT TURN \n";
        timeline += gameOfLife2;
        Approvals.verify(timeline);
    }

    @Test
    void testTwoFirstLevelNeighbours() {
        String timeline = "";
        GameOfLife gameOfLife = new GameOfLife();
        gameOfLife.board.setAlive(4, 4);
        gameOfLife.board.setAlive(3, 5);
        gameOfLife.board.setAlive(4, 6);
        gameOfLife.board.setAlive(5, 5);
        timeline += HexPrinter.print(gameOfLife.board);
        GameOfLife gameOfLife2 = gameOfLife.advanceTurn();
        timeline += "\n NEXT TURN \n";
        timeline += HexPrinter.print(gameOfLife2.board);
        Approvals.verify(timeline);
    }

    @Test
    void testEverything() {
        String fullTimeline = "";

        for (int l1 = 0; l1 <= 6; l1++) {
            for (int l2 = 0; l2 <= 6; l2++) {
                String timeline = advanceBoardWithNeighbours(l1, l2);
                fullTimeline += timeline + "\n\n\n\n";
            }
        }
        Approvals.verify(fullTimeline);
    }

    private String advanceBoardWithNeighbours(int numberOfLevelOneNeighbours, int numberOfLevelTwoNeighbours) {
        String title = String.format("Neighbours(Level 1 , Level 2) = (%s, %s) => \n", numberOfLevelOneNeighbours, numberOfLevelTwoNeighbours);
        GameOfLife gameOfLife = createGameWithNeighbours(numberOfLevelOneNeighbours, numberOfLevelTwoNeighbours, true);
        return printAdvanceBoard(title, gameOfLife);
    }

    private String printAdvanceBoard(String title, GameOfLife gameOfLife) {
        String timeline = "";
        timeline += title;


        timeline += HexPrinter.print(gameOfLife.board);
        timeline += "\n NEXT TURN \n";
        GameOfLife postTurnBoard = gameOfLife.advanceTurn();
        timeline += HexPrinter.print(postTurnBoard.board);
        return timeline;
    }

    @Test
    void testCentreComesAlive() {
        GameOfLife boardWithNeighbours = createGameWithNeighbours(2, 3, false);
        Approvals.verify(printAdvanceBoard("Centre (4,4) comes alive\n", boardWithNeighbours));
    }

    public static GameOfLife createGameWithNeighbours(int numberOfLevelOneNeighbours, int numberOfLevelTwoNeighbours, boolean includeCentre) {
        Cell centre = new Cell(4, 4);
        ArrayList<Cell> cells = new ArrayList<>();

        if (includeCentre) {
            cells.add(centre);
        }

        List<Cell> level1 = centre.getLevelOneNeighbours().subList(0, numberOfLevelOneNeighbours);
        cells.addAll(level1);

        List<Cell> level2 = centre.getLevelTwoNeighbours().subList(0, numberOfLevelTwoNeighbours);
        cells.addAll(level2);

        return new GameOfLife(cells);
    }
}
