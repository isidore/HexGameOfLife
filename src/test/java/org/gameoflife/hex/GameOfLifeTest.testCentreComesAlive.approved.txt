Centre comes alive
| = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = =  |
|    _ _           _ _           _ _           _ _           _ _           |
|  /     \       /     \       /     \       /     \       /     \         |
| /  0,0  \ _ _ /  2,0  \ _ _ /  4,0  \ _ _ /  6,0  \ _ _ /  8,0  \ _ _    |
| \       /     \       /X X X\       /     \       /     \       /     \  |
|  \ _ _ /  1,1  \ _ _ /X  X  X\ _ _ /  5,1  \ _ _ /  7,1  \ _ _ /  9,1  \ |
|  /     \       /     \X  X  X/     \       /     \       /     \       / |
| /  0,2  \ _ _ /  2,2  \X_X_X/  4,2  \ _ _ /  6,2  \ _ _ /  8,2  \ _ _ /  |
| \       /     \       /X X X\       /     \       /     \       /     \  |
|  \ _ _ /  1,3  \ _ _ /X  X  X\ _ _ /  5,3  \ _ _ /  7,3  \ _ _ /  9,3  \ |
|  /     \       /X X X\X  X  X/     \       /     \       /     \       / |
| /  0,4  \ _ _ /X  X  X\X_X_X/  4,4  \ _ _ /  6,4  \ _ _ /  8,4  \ _ _ /  |
| \       /     \X  X  X/X X X\       /     \       /     \       /     \  |
|  \ _ _ /  1,5  \X_X_X/X  X  X\ _ _ /  5,5  \ _ _ /  7,5  \ _ _ /  9,5  \ |
|  /     \       /     \X  X  X/     \       /     \       /     \       / |
| /  0,6  \ _ _ /  2,6  \X_X_X/  4,6  \ _ _ /  6,6  \ _ _ /  8,6  \ _ _ /  |
| \       /     \       /X X X\       /     \       /     \       /     \  |
|  \ _ _ /  1,7  \ _ _ /X  X  X\ _ _ /  5,7  \ _ _ /  7,7  \ _ _ /  9,7  \ |
|  /     \       /     \X  X  X/     \       /     \       /     \       / |
| /  0,8  \ _ _ /  2,8  \X_X_X/  4,8  \ _ _ /  6,8  \ _ _ /  8,8  \ _ _ /  |
| \       /     \       /     \       /     \       /     \       /     \  |
|  \ _ _ /  1,9  \ _ _ /  3,9  \ _ _ /  5,9  \ _ _ /  7,9  \ _ _ /  9,9  \ |
|        \       /     \       /     \       /     \       /     \       / |
|         \ _ _ /       \ _ _ /       \ _ _ /       \ _ _ /       \ _ _ /  |
|                                                                          |
| = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = =  |

 NEXT TURN 
| = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = =  |
|    _ _           _ _           _ _           _ _           _ _           |
|  /     \       /     \       /     \       /     \       /     \         |
| /  0,0  \ _ _ /  2,0  \ _ _ /  4,0  \ _ _ /  6,0  \ _ _ /  8,0  \ _ _    |
| \       /     \       /     \       /     \       /     \       /     \  |
|  \ _ _ /  1,1  \ _ _ /  3,1  \ _ _ /  5,1  \ _ _ /  7,1  \ _ _ /  9,1  \ |
|  /     \       /     \       /X X X\       /     \       /     \       / |
| /  0,2  \ _ _ /  2,2  \ _ _ /X  X  X\ _ _ /  6,2  \ _ _ /  8,2  \ _ _ /  |
| \       /     \       /X X X\X  X  X/     \       /     \       /     \  |
|  \ _ _ /  1,3  \ _ _ /X  X  X\X_X_X/  5,3  \ _ _ /  7,3  \ _ _ /  9,3  \ |
|  /     \       /X X X\X  X  X/X X X\       /     \       /     \       / |
| /  0,4  \ _ _ /X  X  X\X_X_X/X  X  X\ _ _ /  6,4  \ _ _ /  8,4  \ _ _ /  |
| \       /     \X  X  X/X X X\X  X  X/     \       /     \       /     \  |
|  \ _ _ /  1,5  \X_X_X/X  X  X\X_X_X/  5,5  \ _ _ /  7,5  \ _ _ /  9,5  \ |
|  /     \       /     \X  X  X/X X X\       /     \       /     \       / |
| /  0,6  \ _ _ /  2,6  \X_X_X/X  X  X\ _ _ /  6,6  \ _ _ /  8,6  \ _ _ /  |
| \       /     \       /     \X  X  X/     \       /     \       /     \  |
|  \ _ _ /  1,7  \ _ _ /  3,7  \X_X_X/  5,7  \ _ _ /  7,7  \ _ _ /  9,7  \ |
|  /     \       /     \       /     \       /     \       /     \       / |
| /  0,8  \ _ _ /  2,8  \ _ _ /  4,8  \ _ _ /  6,8  \ _ _ /  8,8  \ _ _ /  |
| \       /     \       /     \       /     \       /     \       /     \  |
|  \ _ _ /  1,9  \ _ _ /  3,9  \ _ _ /  5,9  \ _ _ /  7,9  \ _ _ /  9,9  \ |
|        \       /     \       /     \       /     \       /     \       / |
|         \ _ _ /       \ _ _ /       \ _ _ /       \ _ _ /       \ _ _ /  |
|                                                                          |
| = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = =  |
